File routing xml schema

bording opened the pull request:
This adds XML schema validation to the FileBasedDynamicRouting parser. The schema file is loaded as an embedded resource.

I did start out with the schema defined as as string in the class, so it maybe it makes sense to squash some of these commits before merging?

While getting the schema working, I did end up making a tweak to the version currently on the docs page. I've made it require at least once instance if their is an endpoint defined, because otherwise there's no point in the endpoint being defined in the file in the first place.

@timbussmann See anything else you'd want to add to this?

Connects to Particular/PlatformDevelopment#814

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTableParser.cs:
```diff
@@ -1,39 +1,50 @@
     {
+        public FileRoutingTableParser()
+        {
+            var stream = GetType().Assembly.GetManifestResourceStream("NServiceBus.Routing.FileBasedDynamicRouting.endpoints.xsd");
```
shouldn't the stream use a `using`?

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/endpoints.xsd:
```diff
@@ -0,0 +1,21 @@
+        <xs:element maxOccurs="unbounded" name="endpoint">
+          <xs:complexType>
+            <xs:sequence>
+              <xs:element maxOccurs="unbounded" name="instance">
```
would probably make sense to explicitly state that it requires at least one element by specifying `minOccurs="1"` instead of the implicit default value?

timbussmann commented:
looks very similar to my spiked attempt at built-in schema validation. Do we need to catch the schema validation exception and add some further text to it to help users solve their invalid config xml?

SzymonPobiega commented:
:+1: for catching and wrapping the validation exception. I don't see any other issues apart from the ones pointed out by @timbussmann . Good stuff!

bording commented:
I addressed the disposing and feedback on the schema.

I also wrapped the exceptions from from ReloadData, in the general error message that was already being used. I did add the file path to the text.

Any other additional wording in the message that would be helpful?

When those errors are logged, does the log show the InnerException message?

SzymonPobiega commented:
@bording based on my experience with troubleshooting NHibernate persistence inner exceptions are logged as well (at least in the logger adapters I've seen)

SzymonPobiega commented:
I like it. :+1:

timbussmann commented on src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs:
```diff
@@ -20,16 +20,25 @@ public FileRoutingTable(string filePath, TimeSpan checkInterval, IAsyncTimer tim
             this.fileAccess = fileAccess;
             this.maxLoadAttempts = maxLoadAttempts;
+
+            errorMessage = $"Error while reading routing table file at {filePath}.";
```
Maybe something like "An error occured while reading the endpoint instance mapping file at {filePath}. Please check the routing documentation and see the inner exception for more details." ?

bording commented:
I updated the error text to:

`An error occured while reading the endpoint instance mapping file at {filePath}. See the inner exception for more details.`

The branch has been squashed and rebased.
